File: 2024/05/python/main.py
from fileinput import input
from functools import cmp_to_key

silver = 0
gold = 0

lines = [line.strip() for line in input()]
split = lines.index("")

ordering, updates = lines[:split], lines[split + 1 :]


def cmp(x, y):
    if f"{y}|{x}" in ordering:
        return 1
    return -1


for update in updates:
    raw = update.split(",")
    fixed = list(sorted(raw, key=cmp_to_key(cmp)))

    mid = int(fixed[len(fixed) // 2])
    if fixed == raw:
        silver += mid
    else:
        gold += mid

print("silver:", silver)
print("gold:", gold)

File: 2024/05/rust/src/main.rs
#![feature(slice_split_once)]

use std::cmp::Ordering;
use std::collections::HashSet;
use std::io;

fn main() -> io::Result<()> {
    let mut lines = io::stdin().lines().flatten();
    let mut heap = HashSet::<(u16, u16)>::new();

    while let Some(s) = lines.next() {
        if let Some((a, b)) = s.split_once('|') {
            heap.insert((a.parse().unwrap(), b.parse().unwrap()));
        } else {
            break;
        }
    }

    let mut silver: usize = 0;
    let mut gold: usize = 0;

    for line in lines {
        let mut nums: Vec<u16> = line.split(',').map(|n| n.parse().unwrap()).collect();
        let presort = nums.clone();
        let mid = nums.len() / 2;
        nums.sort_unstable_by(|&a, &b| match heap.contains(&(a, b)) {
            true => Ordering::Less,
            false => Ordering::Greater,
        });

        if nums.iter().eq(presort.iter()) {
            silver += presort[mid] as usize;
        } else {
            gold += nums[mid] as usize;
        }
    }

    println!("silver: {silver}");
    println!("gold: {gold}");

    return Ok(());
}
